### PR TITLE
Add Checkin API and Modify Badge Plugin

### PR DESCRIPTION
**Pull request by Sak1012**
This PR adds CheckinAPI to the eventyay-tickets component that can be used by open-event-checkin to perform checkin

Example Request

>POST `http://localhost:8000/api/v1/organizers/admin/checkinrpc/redeem/`

Request body
```json
{
  "secret": "r7zxu7b65p38929a4da25hr6u39e29uu",
  "source_type": "barcode",
  "lists":["1","2"],
  "force": false,
  "ignore_unpaid": false,
  "nonce": "Pvrk50vUzQd0DhdpNRL4I4OcXsvg70uA",
  "datetime": null,
  "questions_supported": false
}
```

Response
```json
{
    "status": "ok",
    "require_attention": false,
    "position": {
        "id": 19,
        "order": "H0ZCF",
        "positionid": 1,
        "item": 3,
        "variation": null,
        "price": "0.00",
        "attendee_name": "Tony Stark",
        "attendee_name_parts": {
            "_scheme": "given_family",
            "given_name": "Tony",
            "family_name": "Stark"
        },
        "company": null,
        "street": null,
        "zipcode": null,
        "city": null,
        "country": null,
        "state": null,
        "attendee_email": null,
        "voucher": null,
        "tax_rate": "0.00",
        "tax_value": "0.00",
        "secret": "r7zxu7b65p38929a4da25hr6u39e29uu",
        "addon_to": null,
        "subevent": null,
        "checkins": [
            {
                "id": 1,
                "datetime": "2024-10-09T13:35:50.705298Z",
                "list": 2,
                "auto_checked_in": false,
                "type": "entry"
            }
        ],
        "downloads": [
            {
                "output": "pdf",
                "url": "http://localhost:8000/api/v1/organizers/admin/events/bc102/orderpositions/19/download/pdf/"
            }
        ],
        "answers": [],
        "tax_rule": null,
        "pseudonymization_id": "P73AWLUZKN",
        "seat": null,
        "require_attention": false,
        "order__status": "p"
    },
    "list": {
        "id": 2,
        "name": "Default",
        "event": "bc102",
        "subevent": null,
        "include_pending": false
    }
}
```

## Summary by Sourcery

Add a new Checkin API to facilitate check-in operations for events, supporting various media types and enhancing the check-in process with new serializers and validation mechanisms.

New Features:
- Introduce a new Checkin API to the eventyay-tickets component, allowing the open-event-checkin application to perform check-ins via a new endpoint.

Enhancements:
- Add support for handling different media types, including barcode and NFC, for check-in processes.
- Implement a new serializer for handling input data for the Checkin RPC Redeem process, ensuring proper validation and context setup.

### CODE REVIEW

File: src/pretix/api/views/checkin.py [hunk 1]
```diff
@@ -188,6 +193,254 @@ def check_rules_qs(self, queryset, name, value):
...
+
+    q = Q(secret=raw_barcode)
+
+    if raw_barcode.isnumeric() and not untrusted_input and legacy_url_support:
+        q |= Q(pk=raw_barcode)
+
+
+    op_candidates = list(queryset.filter(q)) #filtering queryset with q to fetch orderposition
+
+    print('\n',op_candidates,'\n')
```
Comment (hongquan) [inline src/pretix/api/views/checkin.py]: Use logging, don't use print.

File: src/pretix/api/views/checkin.py [hunk 2]
```diff
@@ -188,6 +193,254 @@ def check_rules_qs(self, queryset, name, value):
...
+
+
+
+    q = Q(secret=raw_barcode)
+
+    if raw_barcode.isnumeric() and not untrusted_input and legacy_url_support:
+        q |= Q(pk=raw_barcode)
+
+
+    op_candidates = list(queryset.filter(q)) #filtering queryset with q to fetch orderposition
```
Comment (hongquan) [inline src/pretix/api/views/checkin.py:359]: Don't know why you have to load many rows and then, use only the first item

```py
op = op_candidates[0]
```

You should use `first` or `earliest` method on queryset.

File: src/pretix/api/views/checkin.py [hunk 3]
```diff
@@ -188,6 +193,254 @@ def check_rules_qs(self, queryset, name, value):
...
+        'image/png', 'image/jpeg', 'image/gif', 'application/pdf'
+    )
+    if cf.type not in allowed_types:
+        raise BaseValidationError('The submitted file "{fid}" has a file type that is not allowed in this field.'.format(fid=data))
+    if cf.file.size > settings.FILE_UPLOAD_MAX_SIZE_OTHER:
+        raise BaseValidationError('The submitted file "{fid}" is too large to be used in this field.'.format(fid=data))
+
+    return cf.file
+
+def _redeem_process(*, checkinlists, raw_barcode, answers_data, datetime, force, checkin_type, ignore_unpaid, nonce,
```
Comment (hongquan) [inline src/pretix/api/views/checkin.py:308]: Don't use parameter name which looks like a type in standard lib (`datetime`). It is confusing.

File: src/pretix/api/views/checkin.py [hunk 4]
```diff
@@ -188,6 +193,254 @@
...
```
[the remark this answers is not in the document]
Comment (hongquan) [reply src/pretix/api/views/checkin.py:308]: Please refactor as sourcery-ai suggested.

**Review by hongquan — CHANGES_REQUESTED** [review]

File: src/pretix/base/models/orders.py [hunk 5]
```diff
@@ -2047,6 +2047,12 @@ class Meta:
     def sort_key(self):
         return self.addon_to.positionid if self.addon_to else self.positionid, self.addon_to_id or 0
 
+    @cached_property
+    def require_checkin_attention(self):
+        if self.order.checkin_attention or self.item.checkin_attention or (self.variation_id and self.variation.checkin_attention):
```
Comment (hongquan) [inline src/pretix/base/models/orders.py]: Why not just

```py
return (self.order....)
```

File: src/pretix/api/views/checkin.py (hunk 2 above)
Comment (Sak1012) [reply src/pretix/api/views/checkin.py:359]: I have added a case to check ambiguity if multiple are found

**Review by mariobehling — CHANGES_REQUESTED** [review]
Please fix tests. Except Docker this PR breaks tests that were working before.

File: src/pretix/base/models/orders.py (hunk 5 above)
Comment (Sak1012) [reply src/pretix/base/models/orders.py]: Done 👍

File: src/pretix/api/views/checkin.py (hunk 1 above)
Comment (Sak1012) [reply src/pretix/api/views/checkin.py]: Sorry had it there for testing now it's been removed

**Pull request closed without merging**